### ISSUE TITLE
Feature Request: Auto Generated Components

### ISSUE THREAD

**Issue by marcellogalhardo** [square/anvil#57]
I'm starting to migrate a large project to Anvil and I'm converting each Application Component, from all "Feature Applications" to use `@MergeComponent`. It is working as expected, really good job!

However, I notice that I have to duplicate in each application module the same component definition that looks somehow like this (a little more elaborate to say):

```kotlin
@Singleton
@MergeComponent(
    // I'm using the scope annotation itself here.
    scope = Singleton::class
)
interface AppComponent {

    @Component.Factory
    interface Factory {
        // Simple example, might have more params for other cases...
        fun create(@BindsInstance app: Application): AppComponent
    }
}
```

As the definition is always the same, I wonder if Anvil could have a mechanism to "contribute with a component", or better, auto generate a given component inside a custom module so Dagger can process it from the root and therefore, we can reuse it. For example, something like:

```kotlin
// AppComponent lives inside a `android.library` base module
@GeneratesComponent(
    generateIn = GeneratesAppComponent::class,
    scope = Singleton::class // to allow contributions
)
interface AppComponent {

    @Component.Factory
    interface Factory {
        fun create(@BindsInstance app: Application): AppComponent
    }
}

// Application lives inside an `android.application` root module.
// To use, in any class, you would annotate with the generation marker, and Anvil would generate an `AnvilAppComponent` merge component:
@GeneratesAppComponent // simple annotation
class App: Application(), ComponentOwner<AppComponent> {

    // Basic component accessor for example.
    override val component = Dagger_AnvilAppComponent
        .factory()
        .create(app)
}
```

This doesn't sounds a lot but we have more than 50 Feature Apps and this would be useful to not need to duplicate these components, that are literally the same. Also it allow us to use the same interface, defined in a "bottom module" from the root app module and contribute to it, respecting the same `AppComponent` interface.

TL;DR: I'd like to have to create "predefined components" designed by myself - that will be mount using `@ContributesTo` - that will be auto generated inside the App module.

Does it make sense or it goes out of the scope of this library? 🤔

**Comment by vRallev** [comment]
I'm sorry, but this goes far beyond what Anvil is supposed to do. This is out of scope.
